### ISSUE TITLE
get_cvmix.sh and get_BGC.sh not working

### ISSUE THREAD

**Issue by worleyph** [E3SM-Project/E3SM#3377]
executing the scripts 

     components/mpas-source/src/core_ocean/get_cvmix.sh
     components/mpas-source/src/core_ocean/get_BGC.sh

no longer eliminates the need to download updated copies of CVMIX and BGC during case builds on Cori-KNL. From the e3sm.bldlog ...

      ** Using git to acquire cvmix source. **
      ** Using git to acquire BGC source. **

**Comment by jgfouca** [comment]
@worleyph , this is likely related to the CMake-ification of the MPAS build in E3SM. I didn't quite understand your comment. Are you saying running these scripts once, either by-hand or as part of a build, should prevent them from being run again during subsequent builds of the same case?

**Comment by worleyph** [comment]
By-hand certainly. That was the original motivation. @philipwjones can give you the background to this as Phil wrote the original scripts, I believe.

**Comment by jgfouca** [comment]
@worleyph , just looking at the core_ocean/Makefile vs. ocean.cmake, I can clearly see that I changed this behavior (inadvertently). Sorry about that! Fix incoming.

**Comment by philipwjones** [comment]
@jgfouca - I didn't write the originals, but I did modify them to avoid downloading files for every build. Basically, this is a case where we need to get the libs if they haven't yet been downloaded, but if the correct version is there and the files have been downloaded, the script should detect that and not do another download. I agree that it was something probably missed in the CMake transition.

**Comment by jgfouca** [comment]
@philipwjones,

I took another look at the ocean Makefile, this is the target for getting cvmix:
```
cvmix_source: get_cvmix.sh
        (/bin/bash ./get_cvmix.sh)
        (cd cvmix)
```
Since `cvmix_source` is a "fake" target (not associated with a real file/directory) won't it get run every time the user runs Make?

**Comment by philipwjones** [comment]
@jgfouca - the script gets run every time, but the first part of the script checks for the presence of the proper git clone and avoids a new clone/download. So as long as the script gets run by the CMake version, it should work properly?

**Comment by jgfouca** [comment]
Oh, it looks like the code to avoid repeating downloads is in the script itself, not the build system. In that case, I'm not sure why things aren't working as they were before.

**Comment by philipwjones** [comment]
Hmmm...agreed.

**Comment by jgfouca** [comment]
@worleyph , @philipwjones , when you run the scripts by-hand, how does it know to where to put the code? The build system expects e.g. cvmix to be at a specific location or else it will try to redownload it.

@worleyph , I may need more context as to when and where you ran the download command by hand.

**Comment by worleyph** [comment]
My instructions from the MPAS team were to ...

     cd components/mpas-source/src/core_ocean
     ./get_cvmix.sh
     ./get_BGC.sh

This is been working for a couple of years now.

**Comment by jgfouca** [comment]
Ahhh, so you are downloading them into your **source** tree! Thanks, this is exactly what I need to debug.

**Comment by jgfouca** [comment]
For those interested, a properly configured CMake project will generally not ever touch a users source tree. It does all its work in the build (AKA binary) area. I was following this pattern with CVMIX and BGC, which, after the cmake transition, were being downloaded into the build area. The MPAS Makefile build system is an in-source build system, so it was putting CVMIX and BGC directly into the source tree. This meant the user could download cvmix/bgc into their source manually and not have to repeat those steps during their builds.

It would be pretty easy for me to duplicate this behavior in CMake, but we may need to think about the details a bit. I can think of three options:

1. We violate the out-of-source philosophy of CMake and always download CVMIX/BGC directly into the user's source tree. This would exactly mimic the behavior of the old Makefile system.
2. CMake checks to see if CVMIX/BGC exists in the user's source tree. If so, it uses those; otherwise, it will download them into the *build* tree.
3. We keep the current CMake behavior, treating CVMIX/BGC as build artifacts, and always use CVMIX/BGC from the build tree.

@worleyph , @philipwjones , let me know what I should implement.

**Comment by worleyph** [comment]
The goal is not to require the user to have to enter a github password in the middle of a build every time that the model is built. Running these scripts manually after the initial checkout (and after MPAS updates that require new CVMIX and BGC updates)  was sufficient. I do not care how this is implemented, just that the same goal is achieved.

**Comment by jgfouca** [comment]
Just out of curiousity, why are the dependencies of ocean on cvmix/bgc not implemented as git submodules? It looks like the get_*.sh are basically doing the same thing a git submodule would do.

**Comment by jgfouca** [comment]
@worleyph , I never have to enter my github password. Do you have a passwordless ssh key-pair set up for GitHub?

**Comment by rljacob** [comment]
I've wanted this converted to submodules for a while.

If people want passwords in their github ssh-keys, we should allow that.   I have them and use ssh-agent to avoid entering my password a bunch of times.

**Comment by worleyph** [comment]
> Do you have a passwordless ssh key-pair set up for GitHub?

Not at the computer centers (nersc, alcf, olcf).

**Comment by philipwjones** [comment]
Two things - yes submodule behavior is what these are mimicking - that decision was some time ago and I don't remember why the script approach was chosen. It's possible we could change to a submodule - will need to talk to the other to remind me the reason why they're not.

On the GitHub passwords - while I do have ssh key pairs defined, some systems ask for a password to access your ssh key file, so can still trigger a different password query. does help on most other systems though.

**Comment by jgfouca** [comment]
Thanks @philipwjones , I suppose we will need to know if cvmix/bgc can be submodules before proceeding since that's the preferable (IMO) solution. If not, then we need to decide on which of the three approaches I described above to go with.

**Comment by worleyph** [comment]
If we are not going to fix this immediately, we need to advertise this fact. I wasn't the only user of the "feature". Especially newish users can be caught by the fact that builds will hang if you are not monitoring them, looking for the password request.

**Comment by rljacob** [comment]
@mark-petersen or @jonbob may recall why cvmix/bgc were done that way.

**Comment by jgfouca** [comment]
@worleyph , even if I fixed this right now, it would be days before the fix would be available to E3SM developers because we'd have to wait for the MPAS submodule to get updated.

The workaround is to ensure that the user has passwordless access to GitHub.

I'm not sure how to advertise this to users. The Cmake stuff has been in-place for a couple months on master and you're the first person to notice this change, so maybe (hopefully) not many people were impacted.

**Comment by jonbob** [comment]
@rljacob - I think that decision was made before I joined ACME/E3SM. I could check with Doug Jacobsen...

**Comment by jgfouca** [comment]
One thing would should consider. If we make cvmix/bgc submodules, then E3SM will become a recursive submodule repository. This will have implications to all our users as they will have to remember to add "--recursive" to all their submodule operations.

**Comment by worleyph** [comment]
You’re right. Silly comment on my part. Only takes one hung compile to “advertise” the issue. Let’s try to do it right, whatever that is. I was happy with the original solution, but don’t really care how it is done.

**Comment by jgfouca** [comment]
@douglasjacobsen ^

**Comment by philipwjones** [comment]
Will try to get a consensus here and decide soon. But @jgfouca , if we went the submodule route, it looks like the only change to the CMake build would be to remove the get_* scripts since you're just grabbing all the sources anyway?  And yeah, submodules mean --recursive, but with the increased use of 3rd-party stuff, we may want to get into that habit anyway...

**Comment by jgfouca** [comment]
@philipwjones , we would need to change both the CMake and Makefile build systems. Fortunately, the changes should be super-easy since we're just deleting the special handling that was needed for cvmix/bgc.

**Comment by xylar** [comment]
As I just emailed to LANL folks, I'm good with switching to gitsubmodules for these 2.  I agree, it shouldn't be hard and the reasons for not doing that originally are lost to the dustbin of time.  I don't think there's any need to drag Doug Jacobsen into this.

**Comment by rljacob** [comment]
@jgfouca when you said:  "If we make cvmix/bgc submodules, then E3SM will become a recursive submodule repository."  Did you mean mpas-ocean would become a submodule repo?

**Comment by mark-petersen** [comment]
I'm happy with changing to submodules for BGC and cvmix. I also can't remember why we did not do that originally.

**Comment by douglasjacobsen** [comment]
@jgfouca, @jonbob:
Not that it matters much anymore, but things have changed significantly since we decided to write those scripts. At the time, we were attempting to support multiple usage models of the cvmix and BGC components. Including the use of the svn bridge that github provides. This was primarily to help allow people outside of LANL / DOE (like the NCAR / Academic folks) to checkout and build MPAS-O with these components to test them out.

When I wrote them, I also tried to allow multiple git protocols, and even using wget to download the components if you didn't have git or svn on the machine you were working on. That's because MPAS-O can be downloaded as an archive to be run in stand-alone mode, and we don't really have any control over what method these people use to acquire the MPAS code.

So, instead of using submodules (which were my preference at the time) I wrote these scripts to help minimize the amount of support I would have to provide to people with more awkward system configurations.

@rljacob He means that if cvmix and BGC are converted to be submodules, the MPAS-O component of E3SM (which is already a submodule) will have submodules inside of it. So E3SM will need to recursively fetch submodules. I'm unsure what your confluence checkout commands look like for acquiring submodules, but it's something like `git submodule update --init --recursive`

**Comment by jgfouca** [comment]
Thanks, @douglasjacobsen !

It looks like we are in agreement. Unless anyone objects, I'll go ahead and make a PR to implement this change in MPAS (I'm pretty comfortable with git submodules and both build systems, so I think it makes the most sense for me to do it).

**Comment by worleyph** [comment]
Thanks, @jgfouca .

**Comment by philipwjones** [comment]
Great @jgfouca - happy to let you take care of it.

**Comment by rljacob** [comment]
Where should the submodules live?  mpas-source/src/external ?

**Comment by jgfouca** [comment]
@rljacob they are publicly accessible now, so they don't necessarily have to move.

**Comment by rljacob** [comment]
I mean where will they appear in the source code directory structure?

**Comment by jgfouca** [comment]
@rljacob , oh, sorry. They will live exactly where they get downloaded now during the in-source Makefile build.

**Comment by philipwjones** [comment]
Yeah, they're specific to the ocean and typically treated almost like additional source (esp in CMake build), so probably best to keep them in core_ocean where they sit now.

**Comment by rljacob** [comment]
So it would be in mpas-source/src/core_ocean

**Comment by philipwjones** [comment]
Yup

**Comment by jgfouca** [comment]
https://github.com/MPAS-Dev/MPAS-Model/pull/420

**Comment by jgfouca** [comment]
Fixed by #3507
